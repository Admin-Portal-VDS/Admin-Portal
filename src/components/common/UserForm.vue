<template>
  <div class="form-container">
    <form @submit.prevent="handleSubmit">
      <!-- Basic Info Section -->
      <div class="section">
        <h2>Basic Info</h2>
        <div
          v-for="field in basicInfoFields"
          :key="field.id"
          class="form-group"
        >
          <label :for="field.id">{{ field.label }}</label>
          <input
            :id="field.id"
            v-model="form[field.model]"
            :type="field.type"
            :placeholder="field.placeholder"
            :required="field.required"
            :data-id="field.id"
            class="full-width-input"
          />
        </div>

        <!-- Select Group(s) Dropdown -->
        <div class="form-group">
          <label for="groups">Select Group(s)</label>
          <select
            data-id="groups"
            id="groups"
            v-model="form.groups"
            class="select-group"
          >
            <option value="">Select Group(s)</option>
            <option
              v-for="group in groups"
              :key="group.value"
              :value="group.value"
              :data-id="group.value"
            >
              {{ group.label }}
            </option>
          </select>
        </div>
      </div>

      <div class="divider">
        <span></span>
      </div>

      <!-- Settings and Permissions Section -->
      <div class="section">
        <h2>Settings and Permissions</h2>
        <div class="user-type">
          <select
            data-id="userType"
            id="userType"
            v-model="form.userType"
            required
          >
            <option
              v-for="option in userTypeOptions"
              :key="option.value"
              :value="option.value"
              :data-id="option.value"
            >
              {{ option.text }}
            </option>
          </select>
        </div>
      </div>

      <!-- User Type Section (Moved after Settings and Permissions) -->
      <div class="section">
        <div
          v-for="(label, key) in permissionFields"
          :key="key"
          class="form-group"
        >
          <label>
            <input
              type="checkbox"
              v-model="form.permissions[key]"
              :data-id="`permission-${key}`"
            />
            {{ label }}
          </label>
        </div>
      </div>

      <div class="divider">
        <span></span>
      </div>

      <!-- Extensions Section -->
      <div class="section">
        <h2>Extensions</h2>
        <div
          v-for="(label, key) in extensionOptions"
          :key="key"
          class="form-group"
        >
          <label>
            <input
              type="radio"
              v-model="form.extensionOption"
              :value="key"
              :data-id="`extension-${key}`"
            />
            {{ label }}
          </label>
        </div>

        <div
          data-id="notice"
          v-if="form.extensionOption === 'purchase'"
          class="notice"
        >
          You will be brought to the store to purchase an extension after saving
          the user. Once you have purchased an extension, go to the Extensions
          page to assign the extension to this user.
        </div>
      </div>
    </form>
  </div>
</template>

<script>
export default {
  data() {
    return {
      form: {
        firstName: '',
        lastName: '',
        username: '',
        email: '',
        confirmEmail: '',
        groups: '',
        userType: '',
        permissions: {
          serviceAnnouncements: false,
          webAccess: true,
          sendEmail: true,
          manageCallerId: false,
          manageCallBlocking: false,
        },
        extensionOption: '',
      },
      basicInfoFields: [
        {
          id: 'firstName',
          model: 'firstName',
          type: 'text',
          label: 'First Name*',
          placeholder: 'Enter first name',
          required: true,
        },
        {
          id: 'lastName',
          model: 'lastName',
          type: 'text',
          label: 'Last Name*',
          placeholder: 'Enter last name',
          required: true,
        },
        {
          id: 'username',
          model: 'username',
          type: 'text',
          label: 'Username*',
          placeholder: 'Enter username',
          required: true,
        },
        {
          id: 'email',
          model: 'email',
          type: 'email',
          label: 'Email*',
          placeholder: 'Enter email',
          required: true,
        },
        {
          id: 'confirmEmail',
          model: 'confirmEmail',
          type: 'email',
          label: 'Confirm Email*',
          placeholder: 'Confirm email',
          required: true,
        },
      ],
      groups: [
        { value: 'group1', label: 'Group 1' },
        { value: 'group2', label: 'Group 2' },
        { value: 'group3', label: 'Group 3' },
      ],
      userTypeOptions: [
        { value: '', text: 'Select User Type' },
        { value: 'accountAdmin', text: 'V Account Administrator' },
        { value: 'billingAdmin', text: 'Billing Admin' },
        { value: 'callQueuesAdmin', text: 'Call Queues Admin' },
        {
          value: 'callQueuesReportingAdmin',
          text: 'Call Queues Reporting Admin',
        },
        { value: 'endUser', text: 'End User' },
        { value: 'endUserNoDashboard', text: 'End User No Dashboard' },
        { value: 'reportsAdmin', text: 'Reports Admin' },
        { value: 'usersAdmin', text: 'Users Admin' },
      ],
      permissionFields: {
        serviceAnnouncements:
          'Update this user with important service announcements',
        webAccess: 'Allow web and mobile app access',
        sendEmail: 'Send new user email',
        manageCallerId:
          'Allow this user to set and block the Outbound Caller ID for their extension(s)',
        manageCallBlocking:
          'Allow this user to manage Call Blocking for their extension(s)',
      },
      extensionOptions: {
        assign: 'Assign an available extension',
        purchase: 'Purchase a new extension',
        none: 'User does not need an extension',
      },
    };
  },
  methods: {
    handleSubmit() {
      console.log('Form submitted:', this.form);
    },
  },
};
</script>

<style scoped>
/* Reset margin and padding for the body */
body {
  background-color: #f5f5f5;
  font-family: Arial, sans-serif;
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

.form-container {
  max-width: 800px;
  width: 100%; /* Make it responsive */
  padding: 20px;
  border-radius: 8px;
  margin-left: 400px;
}
.select-group {
  height: 45px;
  width: 180px;
  margin-top: 10px;
}
.section {
  margin-bottom: 20px;
}
.full-width-input {
  width: 500px;
  height: 20px;
  color: black;
  border: 1px solid rgba(9, 8, 8, 0.64);
  margin-bottom: 10px;
  margin-top: 10px;
}

h2 {
  font-size: 1.5rem;
  color: #333;
  margin-top: 25px;
  margin-bottom: 15px;
}

.form-group {
  margin-bottom: 15px;
}
.user-type {
  margin-top: 25px;
  width: 80px;
}

label {
  display: block;
  font-weight: bold;
  margin-bottom: 5px;
}

input,
select {
  padding: 10px;
  font-size: 1rem;
  border: 1px solid #ccc;
  border-radius: 4px;
}

.notice {
  font-size: 0.9rem;
  color: #d9534f;
  background-color: #f9e2e2;
  padding: 10px;
  border: 1px solid #d6b3b3;
  border-radius: 4px;
}
.divider {
  display: flex;
  align-items: center;
  margin: 20px 0;
  color: #888;
}
.divider::before,
.divider::after {
  content: '';
  flex: 1;
  height: 1px;
  background: #ddd;
  margin: 0 0px;
}
/* Responsive Design */
@media (max-width: 768px) {
  .form-container {
    padding: 15px;
  }
}

@media (max-width: 480px) {
  .form-container {
    padding: 10px;
  }
}
</style>
