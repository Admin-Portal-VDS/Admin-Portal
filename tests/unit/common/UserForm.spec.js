// tests/unit/common/UserForm.spec.js
import { mount } from '@vue/test-utils';
import UserForm from '@/components/common/UserForm.vue';
import { useToast } from 'vue-toastification';
import { h, reactive, nextTick } from 'vue';

const mockErrors = reactive({});
const mockFormValues = reactive({
  firstName: '',
  lastName: '',
  username: '',
  email: '',
  confirmEmail: '',
  groups: '',
  userType: '',
  permissionFields: [],
  extensionOption: '',
});

const mockHandleSubmit = jest.fn((fn) => {
  return async (e) => {
    if (e && e.preventDefault) e.preventDefault();
    await fn(mockFormValues);
  };
});
const mockResetForm = jest.fn();
const mockToast = { success: jest.fn() };

jest.mock('vue-toastification');
jest.mock('vee-validate', () => {
  const actual = jest.requireActual('vee-validate');
  return {
    ...actual,
    useForm: jest.fn(() => ({
      handleSubmit: mockHandleSubmit,
      resetForm: mockResetForm,
      errors: mockErrors,
      values: mockFormValues,
    })),
  };
});

useToast.mockReturnValue(mockToast);

const FieldStub = {
  name: 'Field',
  props: ['modelValue'],
  emits: ['update:modelValue'],
  template: `
    <input 
      :value="modelValue" 
      @input="$emit('update:modelValue', $event.target.value)" 
      v-bind="$attrs" 
    />
  `,
};

const ErrorMessageStub = {
  name: 'ErrorMessage',
  props: ['name'],
  setup(props) {
    return () =>
      mockErrors[props.name]
        ? h('div', { class: 'error-msg' }, mockErrors[props.name])
        : null;
  },
};

describe('UserForm.vue', () => {
  let wrapper;

  beforeEach(() => {
    Object.keys(mockErrors).forEach((key) => delete mockErrors[key]);
    Object.assign(mockFormValues, {
      firstName: '',
      lastName: '',
      username: '',
      email: '',
      confirmEmail: '',
      groups: '',
      userType: '',
      extensionOption: '',
    });

    mockHandleSubmit.mockClear();
    mockResetForm.mockClear();
    mockToast.success.mockClear();

    wrapper = mount(UserForm, {
      global: {
        stubs: {
          Field: FieldStub,
          ErrorMessage: ErrorMessageStub,
        },
      },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('shows validation errors when required fields are empty', async () => {
    await wrapper.find('form').trigger('submit');
    Object.assign(mockErrors, {
      firstName: 'First name is required',
      email: 'Email is required',
    });
    await nextTick();
    expect(wrapper.find('[data-id="firstName"] + .error-msg').text()).toContain(
      'First name is required',
    );
    expect(wrapper.find('[data-id="email"] + .error-msg').text()).toContain(
      'Email is required',
    );
  });

  it('validates email format and confirmation', async () => {
    const emailInput = wrapper.find('[data-id="email"]');
    const confirmEmailInput = wrapper.find('[data-id="confirmEmail"]');

    await emailInput.setValue('invalid-email');
    await confirmEmailInput.setValue('invalid-email');
    await wrapper.find('form').trigger('submit');
    Object.assign(mockErrors, { email: 'Email is not valid' });
    await nextTick();
    expect(wrapper.find('[data-id="email"] + .error-msg').text()).toContain(
      'Email is not valid',
    );

    await emailInput.setValue('valid@example.com');
    await confirmEmailInput.setValue('different@example.com');
    await wrapper.find('form').trigger('submit');
    Object.assign(mockErrors, { confirmEmail: 'Email must match' });
    await nextTick();
    expect(
      wrapper.find('[data-id="confirmEmail"] + .error-msg').text(),
    ).toContain('Email must match');
  });

  it('requires at least one permission to be selected', async () => {
    await wrapper.find('form').trigger('submit');
    Object.assign(mockErrors, {
      permissions: 'Please select at least one permission',
    });
    await nextTick();
    expect(wrapper.find('[data-id="error-permissions"]').text()).toContain(
      'Please select at least one permission',
    );
  });

  it('shows purchase notice when extension option is purchase', async () => {
    await wrapper.find('[data-id="extension-purchase"]').setValue(true);
    mockFormValues.extensionOption = 'purchase';
    await nextTick();
    expect(wrapper.find('[data-id="notice"]').exists()).toBe(true);
    expect(wrapper.find('[data-id="notice"]').text()).toMatch(
      'You will be brought to the store to purchase an extension after saving the user. Once you have purchased an extension, go to the Extensions page to assign the extension to this user.',
    );
  });
  it('matches snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
