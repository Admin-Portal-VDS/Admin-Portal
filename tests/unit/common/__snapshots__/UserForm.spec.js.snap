// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserForm.vue matches snapshot 1`] = `
<div class="form-header" data-id="form-header">
  <div class="add-users">Add User</div>
  <div class="cancel"><button class="simple-button">Cancel</button></div>
</div>
<div class="form-container">
  <form>
    <!-- Basic Info Section -->
    <div class="section">
      <h2>Basic Info</h2>
      <div class="form-group"><label for="firstName">First Name*</label><input id="firstName" name="firstName" type="text" placeholder="Enter first name" data-id="firstName" class="full-width-input">
        <!---->
      </div>
      <div class="form-group"><label for="lastName">Last Name*</label><input id="lastName" name="lastName" type="text" placeholder="Enter last name" data-id="lastName" class="full-width-input">
        <!---->
      </div>
      <div class="form-group"><label for="username">Username*</label><input id="username" name="username" type="text" placeholder="Enter username" data-id="username" class="full-width-input">
        <!---->
      </div>
      <div class="form-group"><label for="email">Email*</label><input id="email" name="email" type="email" placeholder="Enter email" data-id="email" class="full-width-input">
        <!---->
      </div>
      <div class="form-group"><label for="confirmEmail">Confirm Email*</label><input id="confirmEmail" name="confirmEmail" type="email" placeholder="Confirm email" data-id="confirmEmail" class="full-width-input">
        <!---->
      </div>
      <div class="form-group"><label for="password">Password*</label><input id="password" name="password" type="password" placeholder="Enter password for user" data-id="password" class="full-width-input">
        <!---->
      </div>
      <div class="form-group"><label for="confirmPassword">Confirm Password*</label><input id="confirmPassword" name="confirmPassword" type="password" placeholder="Enter confirm password for user" data-id="confirmPassword" class="full-width-input">
        <!---->
      </div><!-- Select Group(s) Dropdown -->
      <div class="form-group"><label for="groups">Select Group(s)</label><input as="select" name="groups" data-id="groups" id="groups" class="select-group">
        <!---->
      </div>
    </div>
    <div class="divider"><span></span></div><!-- Settings and Permissions Section -->
    <div class="section">
      <h2>Settings and Permissions</h2>
      <div class="user-type"><input as="select" name="userType" data-id="userType" id="userType"></div>
      <!---->
    </div>
    <div class="section">
      <div class="form-group"><label><input name="permissions" type="checkbox" data-id="permission-serviceAnnouncements" value="serviceAnnouncements"> Update this user with important service announcements</label></div>
      <div class="form-group"><label><input name="permissions" type="checkbox" data-id="permission-webAccess" value="webAccess"> Allow web and mobile app access</label></div>
      <div class="form-group"><label><input name="permissions" type="checkbox" data-id="permission-sendEmail" value="sendEmail"> Send new user email</label></div>
      <div class="form-group"><label><input name="permissions" type="checkbox" data-id="permission-manageCallerId" value="manageCallerId"> Allow this user to set and block the Outbound Caller ID for their extension(s)</label></div>
      <div class="form-group"><label><input name="permissions" type="checkbox" data-id="permission-manageCallBlocking" value="manageCallBlocking"> Allow this user to manage Call Blocking for their extension(s)</label></div>
      <!---->
    </div>
    <div class="divider"><span></span></div><!-- Extensions Section -->
    <div class="section">
      <h2>Extensions</h2>
      <div class="form-group"><label><input type="radio" name="extensionOption" data-id="extension-assign" value="assign"> Assign an available extension</label></div>
      <div class="form-group"><label><input type="radio" name="extensionOption" data-id="extension-purchase" value="purchase"> Purchase a new extension</label></div>
      <div class="form-group"><label><input type="radio" name="extensionOption" data-id="extension-none" value="none"> User does not need an extension</label></div>
      <!---->
    </div>
    <!--v-if--><button type="submit" class="submit-btn">Submit</button>
  </form>
</div>
`;
